server/job-workers/shared-memory-manager.cpp

```diff
@@ -0,0 +1,189 @@
+};
+
+void SharedMemoryManager::init() noexcept {
+  static_assert(sizeof(ControlBlock) % 8 == 0, "check yourself");
+  assert(!control_block_);
+
+  const size_t processes = vk::singleton<job_workers::JobWorkersContext>::get().job_workers_num + std::max(workers_n, 1);
+  if (!memory_limit_) {
+    memory_limit_ = processes * JOB_DEFAULT_MEMORY_LIMIT_PROCESS_MULTIPLIER + sizeof(ControlBlock);
+  }
+  auto *raw_mem = static_cast<uint8_t *>(mmap(nullptr, memory_limit_,
+                                              PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANONYMOUS, -1, 0));
+  assert(raw_mem);
+  const size_t left_memory = memory_limit_ - sizeof(ControlBlock);
+  const uint32_t messages_count = std::min(JOB_SHARED_MESSAGES_COUNT_PROCESS_MULTIPLIER * processes, left_memory / sizeof(JobSharedMessage));
+  control_block_ = new(raw_mem) ControlBlock{};
+  raw_mem += sizeof(ControlBlock);
+  for (uint32_t i = 0; i != messages_count; ++i) {
+    freelist_put(&control_block_->free_messages, raw_mem);
+    raw_mem += sizeof(JobSharedMessage);
+  }
+
+  std::array<memory_resource::extra_memory_raw_bucket, JOB_EXTRA_MEMORY_BUFFER_BUCKETS> extra_memory;
+  control_block_->stats.unused_memory = memory_resource::distribute_memory(extra_memory, processes, raw_mem,
+                                                                           left_memory - sizeof(JobSharedMessage) * messages_count);
+  for (size_t i = 0; i != extra_memory.size(); ++i) {
+    auto &free_extra_buffers = control_block_->free_extra_memory[i];
+    auto &extra_buffers = extra_memory[i];
+    for (size_t mem_index = 0; mem_index != extra_buffers.buffers_count(); ++mem_index) {
+      freelist_put(&free_extra_buffers, extra_buffers.get_extra_pool_raw(mem_index));
+    }
+    control_block_->stats.extra_memory[i].count = extra_buffers.buffers_count();
+  }
+
+  control_block_->stats.memory_limit = memory_limit_;
+  control_block_->stats.messages.count = messages_count;
+}
+
+bool SharedMemoryManager::set_memory_limit(size_t memory_limit) noexcept {
+  if (memory_limit < 3 * sizeof(JobSharedMessage) + sizeof(ControlBlock)) {
+    return false;
+  }
+  memory_limit_ = memory_limit;
+  return true;
+}
+
+JobSharedMessage *SharedMemoryManager::acquire_shared_message() noexcept {
+  assert(control_block_);
+  dl::CriticalSectionGuard critical_section;
+  if (void *free_mem = freelist_get(&control_block_->free_messages)) {
+    auto *message = new(free_mem) JobSharedMessage{};
+    control_block_->workers_table[logname_id].attach(message);
+    ++control_block_->stats.messages.acquired;
+    return message;
+  }
+  ++control_block_->stats.messages.acquire_fails;
+  return nullptr;
+}
+
+void SharedMemoryManager::release_shared_message(JobSharedMessage *message) noexcept {
+  dl::CriticalSectionGuard critical_section;
+  control_block_->workers_table[logname_id].detach(message);
+  if (--message->owners_counter == 0) {
```

Comment:
`message->owners_counter` is always 1 or 0. I think it would be more logical to consider it as bool flag `message->released` and set it to `false` here instead of decrement

Reply:
This is a preparation for immutable memory: for the immutable memory `owners_counter` can be > 1.